jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -3439,31 +3440,62 @@ public static MethodHandle whileLoop(MethodHandle initHandle, MethodHandle predH
 	 */
 	public static MethodHandle doWhileLoop(MethodHandle initHandle, MethodHandle bodyHandle, MethodHandle predHandle) throws NullPointerException, IllegalArgumentException {
 		
-		/* Check the validity of the predicate (must be non-null) and body handle.
-		 * Note: The constraints of initializer and predicate at other aspects will be checked
+		/* Check the validity of the predicate (must be non-null) and body handle. Also
+		 * verify that handles are effectively identical and constrained by body. Note:
+		 * The constraints of initializer and predicate at other aspects will be checked
 		 * later in the generic loop when adding clauses.
 		 */
-		MethodHandle finiHandle = validateArgumentsOfWhileLoop(predHandle, bodyHandle);
+		MethodHandle finiHandle = validateArgumentsOfWhileLoop(initHandle, predHandle, bodyHandle);
 		
 		/* Create the clause array with all these handles to be used by the generic loop */
 		MethodHandle[] handleClause = {initHandle, bodyHandle, predHandle, finiHandle};
 		return loop(handleClause);
 	}
 	
 	/* Validate the body handle against the constraints of a while/do-while loop and generate the 'fini' handle from it */
-	private static MethodHandle validateArgumentsOfWhileLoop(MethodHandle predHandle, MethodHandle bodyHandle) throws NullPointerException, IllegalArgumentException {
+	private static MethodHandle validateArgumentsOfWhileLoop(MethodHandle initHandle, MethodHandle predHandle, MethodHandle bodyHandle) throws NullPointerException, IllegalArgumentException {
 		if ((null == predHandle) || (null == bodyHandle)) {
 			/*[MSG "K065C", "Both The predicate and the loop body must be non-null"]*/
 			throw new NullPointerException(com.ibm.oti.util.Msg.getString("K065C")); //$NON-NLS-1$
 		}
 		
 		MethodType bodyType = bodyHandle.type;
 		Class<?> bodyReturnType = bodyType.returnType;
+		int bodyParamLength = bodyType.parameterCount();
 		
 		/* The signature of the body handle must be either (V, A...)V or (A...)void */
-		if ((void.class != bodyReturnType) && (bodyReturnType != bodyType.arguments[0])) {
-			/*[MSG "K065D", "The return type of loop body: {0} does not match the type of the first argument: {1}"]*/
-			throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065D", bodyReturnType.getName(), bodyType.arguments[0].getName())); //$NON-NLS-1$
+		int bodyIterationVarLength = 0;
+		if (void.class != bodyReturnType) {
+			if (hasNoArgs(bodyType)) {
+				/*[MSG "K065D3", "loop body has a non-void return type and must contain at least one argument. First argument must be {0}"]*/
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065D3", bodyReturnType)); //$NON-NLS-1$
+			}
+			
+			if (bodyReturnType != bodyType.arguments[0]) {
+				MethodType correctType = bodyType.insertParameterTypes(0, bodyReturnType);
+				
+				/*[MSG "K065D2", "The return type of loop body does not match the type of the first argument: {0} != {1}"]*/
```

Comment:
K065D2:  Suggest to only print out the return type and the 1st argument instead of the whole signature; otherwise, it would be confusing to people what it is really comparing here.